### PR TITLE
screenOrientation support for TWAs

### PR DESCRIPTION
**Pull request by sclimwi**
The previous PR (#122 ) is having too many unnecessary files so I made a new branch with the implementation of screenOrientation support.

This one should be clean and right to get reviews. PTAL @PEConn @andreban @glennhartmann @EiraGe

Thanks very much in advance!

### CODE REVIEW

**Comment by sclimwi** [comment]
Hey, All. 

I modified the twa-demo to test on a simple website with res from twa-basic, instead of relying the old twa-notification-delegation files.

It has a conflict with the alpha but I think screen orientation locking has to rely on this `androidx.browser:browser:1.3.0-alpha05`.

PTAL and let me know if there are issues to be taken care of, @PEConn @andreban @glennhartmann @EiraGe.

File: build.gradle [hunk 1]
```diff
@@ -34,6 +33,7 @@ buildscript {
 
 allprojects {
     repositories {
+        maven { url '/home/scli/more-android-sdk/extra/m2repository' }
```
Comment (PEConn) [inline build.gradle:36]: What's this about?

File: demos/twa-basic/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -49,6 +50,10 @@
             <meta-data android:name="android.support.customtabs.trusted.DISPLAY_MODE"
                 android:value="immersive"/>
 
+
```
Comment (PEConn) [inline demos/twa-basic/src/main/AndroidManifest.xml:53]: nit: Random newline.

File: demos/twa-basic/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -49,6 +50,10 @@
...
+            <meta-data android:name="android.support.customtabs.trusted.SCREEN_ORIENTATION"
+                android:value="landscape"/>
```
Comment (PEConn) [inline demos/twa-basic/src/main/AndroidManifest.xml:55]: So I do like the idea of having an example for how to set the value in our basic demo, but I don't think that I want the `twa-basic` demo to launch in landscape. Could you just change this to portrait please?

File: demos/twa-orientation/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+                android:resource="@drawable/splash"/>
+
+            <meta-data android:name="android.support.customtabs.trusted.SPLASH_SCREEN_BACKGROUND_COLOR"
+                android:resource="@color/backgroundColor"/>
+
+            <meta-data android:name="android.support.customtabs.trusted.SPLASH_SCREEN_FADE_OUT_DURATION"
+                android:value="300"/>
+
+            <meta-data android:name="android.support.customtabs.trusted.FILE_PROVIDER_AUTHORITY"
+                android:value="@string/provider_authority"/>
```
Comment (PEConn) [inline demos/twa-orientation/src/main/AndroidManifest.xml:60]: So in this block, pretty much only the default url and the screen orientation meta-data are necessary. If you leave the rest out, you'll lose the splash screen and end up with default colours.

I don't mind leaving the rest in (to make the demo look pretty) or taking them out to make it obvious which are the useful bits of the code. Up to you.

File: build.gradle (hunk 1 above)
Comment (sclimwi) [reply build.gradle:36]: That is the dependency on my local build. Will remove it shortly.

File: demos/twa-basic/src/main/AndroidManifest.xml (hunk 2 above)
Comment (sclimwi) [reply demos/twa-basic/src/main/AndroidManifest.xml:53]: Ack.

File: demos/twa-basic/src/main/AndroidManifest.xml (hunk 3 above)
Comment (sclimwi) [reply demos/twa-basic/src/main/AndroidManifest.xml:55]: Sure.

File: demos/twa-orientation/src/main/AndroidManifest.xml (hunk 4 above)
Comment (sclimwi) [reply demos/twa-orientation/src/main/AndroidManifest.xml:60]: I think for the testing purpose, moving them out would be good. I will remove these lines.

**Comment by sclimwi** [comment]
Hey, @PEConn and @andreban. I resolved the conflicts. Could you please merge this PR.

**Comment by humiston** [comment]
Is there any time estimation for when this functionality will be available via release?

**Comment by andreban** [comment]
@humiston this change depends on an alpha version of `androidx.browser` to become stable. We expect that to happen in a couple of months.

**Comment by MrNeilRobbins** [comment]
Looking forward to this feature. Glad to see it's being worked on. Our TWA needs to be locked to landscape.

I've cloned the repo and tested the demo in the 'alpha' branch BUT the lock doesn't seem to work? When I rotate my device the app rotates. The demo app points to HTML that has a button to enable full screen and lock orientation with JavaScript and that works. But the TWA orientation lock doesn't work. Is there something I'm missing when running the demo.

Glad to help debug.

**Comment by sclimwi** [comment]
Hey Neil,

In order to lock the orientation, the app has to be in 1) TWA mode (the
search bar is not shown), and 2) we have to rely on the chromium build that
has the orientation support code added. I think these are the two most
important pieces in order to lock the orientation.

The demo app there is to show the locking through JS will take precedence
over the default orientation we added.

Best
SL

**Comment by EiraGe** [comment]
@MrNeilRobbins Thanks for testing it. As @sclimwi mentioned, you need to download Chrome Canary and set it as default browser to use the feature. (The code is in Chrome M86 and should roll out to Chrome stable around Mid Oct)

**Comment by MrNeilRobbins** [comment]
Thanks. 

I've installed Chrome Canary on my device (Chrome Canary 86.0.4232.0) OS Android 10; SM-T720 Build/QP1A.190711.020

Then went to:
`Settings > Apps > ... > Default Apps > Browser app > set to Chrome Canary`

But when run **twa-orientation** and it launches on the device the app still rotates freely without locking the orientation.

**Comment by sclimwi** [comment]
@MrNeilRobbins Thanks for checking out that. I do not know if you run through this step yet but just in case:

> Enable debug mode
> Open Chrome on the development device, navigate to chrome://flags, search for an item called Enable command line on non-rooted devices and change it to ENABLED and then restart the browser.
> Next, on the Terminal application of your operating system, use the Android Debug Bridge (installed with Android Studio), and run the following command:
> adb shell "echo '_ --disable-digital-asset-link-verification-for-url=\"https://airhorner.com\"' > /data/local/tmp/chrome-command-line"
> 
> Close Chrome and re-launch your application from Android Studio. The application should now be shown in full-screen.

from this document: https://developers.google.com/web/android/trusted-web-activity/integration-guide. 

This step is to make sure the website is in TWA mode. You may want to change the command to be 

`adb shell "echo '_ --disable-digital-asset-link-verification-for-url=\"https://tulip-forest-limpet.glitch.me\"' > /data/local/tmp/chrome-command-line" `

One more note is that sometime even we set the default browser, the app will still be launched with default Chrome version. We want to make sure we could pick up the right Chrome version. 

Hopefully with the above steps, you can get the orientation up and running.

**Comment by shunxing** [comment]
I tried to set the value to portrait in my manifest but it ultimately doesn't work. 
image
When rotating my simulator, the app still rotates. Can someone explain me whether I did something wrong ? @sclimwi @PEConn

**Pull request merged**